File: model/src/main/java/com/bnorm/barkeep/model/Ingredient.java
package com.bnorm.barkeep.model;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonView;

import javax.annotation.Nullable;

import java.time.Instant;
import java.util.Comparator;

@JsonInclude(JsonInclude.Include.NON_EMPTY)
public interface Ingredient extends HasId, Comparable<Ingredient> {

  @JsonView({Ingredient.class, Recipe.class})
  @Nullable
  String getTitle();

  @JsonView(Ingredient.class)
  @Nullable
  Ingredient getParent();

  @Override
  default int compareTo(Ingredient o) {
    return COMPARATOR.compare(this, o);
  }
}

File: model/src/main/kotlin/com/bnorm/barkeep/model/Ingredient.kt
// Copyright 2017 (C) BNORM Software. All rights reserved.
package com.bnorm.barkeep.model

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonView

interface IngredientSpec {

  @get:JsonView(Ingredient::class, Recipe::class)
  val title: String?

  @get:JsonView(Ingredient::class)
  @get:JsonIgnoreProperties("parent")
  val parent: Ingredient?
}

@JsonInclude(JsonInclude.Include.NON_EMPTY)
interface Ingredient : IngredientSpec, HasId, Comparable<Ingredient> {

  override fun compareTo(other: Ingredient): Int {
    return HasId.COMPARATOR.compare(this, other)
  }
}
